File: src/test/java/me/saro/jwt/java/PerformanceTest.java
//package me.saro.jwt.java;
//
//import me.saro.jwt.old.Jwt;
//import me.saro.jwt.old.JwtAlgorithm;
//import me.saro.jwt.old.JwtKey;
//import me.saro.jwt.old.JwtNode;
//import org.junit.jupiter.api.Assertions;
//import org.junit.jupiter.api.DisplayName;
//import org.junit.jupiter.api.Test;
//
//@DisplayName("[Java] Performance Test")
//public class PerformanceTest {
//
//    @Test
//    @DisplayName("[Java] ES dynamic keys test")
//    public void d_es() {
//        dynamicKeys(100, Jwt.ES256, Jwt.ES384, Jwt.ES512);
//    }
//
//    @Test
//    @DisplayName("[Java] HS dynamic keys test")
//    public void d_hs() {
//        dynamicKeys(1000, Jwt.HS256, Jwt.HS384, Jwt.HS512);
//    }
//
//    @Test
//    @DisplayName("[Java] PS dynamic keys test")
//    public void d_ps() {
//        dynamicKeys(10, Jwt.PS256, Jwt.PS384, Jwt.PS512);
//    }
//
//    @Test
//    @DisplayName("[Java] RS dynamic keys test")
//    public void d_rs() {
//        dynamicKeys(10, Jwt.RS256, Jwt.RS384, Jwt.RS512);
//    }
//
//    @Test
//    @DisplayName("[Java] ES fixed keys test")
//    public void f_es() {
//        fixedKeys(1000, Jwt.ES256, Jwt.ES384, Jwt.ES512);
//    }
//
//    @Test
//    @DisplayName("[Java] HS fixed keys test")
//    public void f_hs() {
//        fixedKeys(5000, Jwt.HS256, Jwt.HS384, Jwt.HS512);
//    }
//
//    @Test
//    @DisplayName("[Java] PS fixed keys test")
//    public void f_ps() {
//        fixedKeys(300, Jwt.PS256, Jwt.PS384, Jwt.PS512);
//    }
//
//    @Test
//    @DisplayName("[Java] RS fixed keys test")
//    public void f_rs() {
//        fixedKeys(500, Jwt.RS256, Jwt.RS384, Jwt.RS512);
//    }
//
//    private void dynamicKeys(int loop, JwtAlgorithm... algs) {
//        for (JwtAlgorithm alg : algs) {
//            long start = System.currentTimeMillis();
//            for (int i = 0 ; i < loop ; i++) {
//                JwtKey key = alg.newRandomKey();
//                var jwt = Jwt.createJwt(key)
//                        .subject("1234567890")
//                        .claim("name", "John Doe")
//                        .claim("admin", true)
//                        .claim("iat", 1516239022)
//                        .build();
//
//                JwtNode node = Assertions.assertDoesNotThrow(() -> Jwt.parseJwt(jwt, nodeKey -> key));
//                Assertions.assertEquals("1234567890", node.getSubject());
//                Assertions.assertEquals("John Doe", node.claimString("name"));
//                Assertions.assertEquals(true, node.claimBoolean("admin"));
//                Assertions.assertEquals(1516239022, node.claimInt("iat"));
//            }
//            System.out.println(alg.getAlgorithmFullName() + " dynamic keys " + loop + " loops time: " + (System.currentTimeMillis() - start) + "ms");
//        }
//    }
//
//    private void fixedKeys(int loop, JwtAlgorithm... algs) {
//        for (JwtAlgorithm alg : algs) {
//            JwtKey key = alg.newRandomKey();
//            long start = System.currentTimeMillis();
//            for (int i = 0 ; i < loop ; i++) {
//                var jwt = Jwt.createJwt(key)
//                        .subject("1234567890")
//                        .claim("name", "John Doe")
//                        .claim("admin", true)
//                        .claim("iat", 1516239022)
//                        .build();
//
//                JwtNode node = Assertions.assertDoesNotThrow(() -> Jwt.parseJwt(jwt, nodeKey -> key));
//                Assertions.assertEquals("1234567890", node.getSubject());
//                Assertions.assertEquals("John Doe", node.claimString("name"));
//                Assertions.assertEquals(true, node.claimBoolean("admin"));
//                Assertions.assertEquals(1516239022, node.claimInt("iat"));
//            }
//            System.out.println(alg.getAlgorithmFullName() + " fixed keys " + loop + " loops time: " + (System.currentTimeMillis() - start) + "ms");
//        }
//    }
//}

File: src/test/kotlin/me/saro/jwt/kotlin/PerformanceTest.kt
package me.saro.jwt.kotlin

import me.saro.jwt.key.JwtAlgorithm
import me.saro.jwt.key.JwtKey.Companion.generateHash
import me.saro.jwt.key.JwtKey.Companion.generateKeyPair
import me.saro.jwt.node.Jwt.Companion.builder
import me.saro.jwt.node.Jwt.Companion.parseOrThrow
import me.saro.jwt.node.JwtNode
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.ThrowingSupplier

@DisplayName("[Kotlin] Performance Test")
class PerformanceTest {
    @Test
    @DisplayName("[Kotlin] ES dynamic keys test")
    fun d_es() {
        dynamicKeysPair(100, JwtAlgorithm.ES256, JwtAlgorithm.ES384, JwtAlgorithm.ES512)
    }

    @Test
    @DisplayName("[Kotlin] HS dynamic keys test")
    fun d_hs() {
        dynamicKeysHash(1000, JwtAlgorithm.HS256, JwtAlgorithm.HS384, JwtAlgorithm.HS512)
    }

    @Test
    @DisplayName("[Kotlin] PS dynamic keys test")
    fun d_ps() {
        dynamicKeysPair(10, JwtAlgorithm.PS256, JwtAlgorithm.PS384, JwtAlgorithm.PS512)
    }

    @Test
    @DisplayName("[Kotlin] RS dynamic keys test")
    fun d_rs() {
        dynamicKeysPair(10, JwtAlgorithm.RS256, JwtAlgorithm.RS384, JwtAlgorithm.RS512)
    }

    @Test
    @DisplayName("[Kotlin] ES fixed keys test")
    fun f_es() {
        fixedKeysPair(1000, JwtAlgorithm.ES256, JwtAlgorithm.ES384, JwtAlgorithm.ES512)
    }

    @Test
    @DisplayName("[Kotlin] HS fixed keys test")
    fun f_hs() {
        fixedKeysHash(5000, JwtAlgorithm.HS256, JwtAlgorithm.HS384, JwtAlgorithm.HS512)
    }

    @Test
    @DisplayName("[Kotlin] PS fixed keys test")
    fun f_ps() {
        fixedKeysPair(300, JwtAlgorithm.PS256, JwtAlgorithm.PS384, JwtAlgorithm.PS512)
    }

    @Test
    @DisplayName("[Kotlin] RS fixed keys test")
    fun f_rs() {
        fixedKeysPair(500, JwtAlgorithm.RS256, JwtAlgorithm.RS384, JwtAlgorithm.RS512)
    }

    private fun dynamicKeysHash(loop: Int, vararg algs: JwtAlgorithm) {
        for (alg in algs) {
            val start = System.currentTimeMillis()
            for (i in 0..<loop) {
                val key = generateHash(alg, 32)
                val jwt = builder()
                    .subject("1234567890")
                    .claim("name", "John Doe")
                    .claim("admin", true)
                    .claim("iat", 1516239022)
                    .build(key)

                val node = Assertions.assertDoesNotThrow<JwtNode>(ThrowingSupplier { parseOrThrow(jwt) })
                Assertions.assertTrue(node.verify(key))
                Assertions.assertEquals("1234567890", node.subject)
                Assertions.assertEquals("John Doe", node.claimString("name"))
                Assertions.assertEquals(true, node.claimBoolean("admin"))
                Assertions.assertEquals(1516239022, node.claimInt("iat"))
            }
            println(alg.name + " dynamic keys " + loop + " loops time: " + (System.currentTimeMillis() - start) + "ms")
        }
    }

    private fun fixedKeysHash(loop: Int, vararg algs: JwtAlgorithm) {
        for (alg in algs) {
            val key = generateHash(alg, 32)
            val start = System.currentTimeMillis()
            for (i in 0..<loop) {
                val jwt = builder()
                    .subject("1234567890")
                    .claim("name", "John Doe")
                    .claim("admin", true)
                    .claim("iat", 1516239022)
                    .build(key)

                val node = Assertions.assertDoesNotThrow<JwtNode>(ThrowingSupplier { parseOrThrow(jwt) })
                Assertions.assertTrue(node.verify(key))
                Assertions.assertEquals("1234567890", node.subject)
                Assertions.assertEquals("John Doe", node.claimString("name"))
                Assertions.assertEquals(true, node.claimBoolean("admin"))
                Assertions.assertEquals(1516239022, node.claimInt("iat"))
            }
            println(alg.name + " fixed keys " + loop + " loops time: " + (System.currentTimeMillis() - start) + "ms")
        }
    }

    private fun dynamicKeysPair(loop: Int, vararg algs: JwtAlgorithm) {
        for (alg in algs) {
            val start = System.currentTimeMillis()
            for (i in 0..<loop) {
                val key = generateKeyPair(alg)
                val jwt = builder()
                    .subject("1234567890")
                    .claim("name", "John Doe")
                    .claim("admin", true)
                    .claim("iat", 1516239022)
                    .build(key.private)

                val node = Assertions.assertDoesNotThrow<JwtNode>(ThrowingSupplier { parseOrThrow(jwt) })
                Assertions.assertTrue(node.verify(key.public))
                Assertions.assertEquals("1234567890", node.subject)
                Assertions.assertEquals("John Doe", node.claimString("name"))
                Assertions.assertEquals(true, node.claimBoolean("admin"))
                Assertions.assertEquals(1516239022, node.claimInt("iat"))
            }
            println(alg.name + " dynamic keys " + loop + " loops time: " + (System.currentTimeMillis() - start) + "ms")
        }
    }

    private fun fixedKeysPair(loop: Int, vararg algs: JwtAlgorithm) {
        for (alg in algs) {
            val key = generateKeyPair(alg)
            val start = System.currentTimeMillis()
            for (i in 0..<loop) {
                val jwt = builder()
                    .subject("1234567890")
                    .claim("name", "John Doe")
                    .claim("admin", true)
                    .claim("iat", 1516239022)
                    .build(key.private)

                val node = Assertions.assertDoesNotThrow<JwtNode>(ThrowingSupplier { parseOrThrow(jwt) })
                Assertions.assertTrue(node.verify(key.public))
                Assertions.assertEquals("1234567890", node.subject)
                Assertions.assertEquals("John Doe", node.claimString("name"))
                Assertions.assertEquals(true, node.claimBoolean("admin"))
                Assertions.assertEquals(1516239022, node.claimInt("iat"))
            }
            println(alg.name + " fixed keys " + loop + " loops time: " + (System.currentTimeMillis() - start) + "ms")
        }
    }
}
